### ISSUE TITLE
[Bug]: Breaking DApps - Unchecked runtime.lastError: Could not establish connection. Receiving end does not exist.

### ISSUE THREAD

**Issue by hamzasharif** [MetaMask/metamask-extension#13465]
### Describe the bug

Hey there,

I've noticed every time I close Chrome and re-open then go to a page that interacts with the MetaMask extension it breaks the sites integrations.

See the following console output
<img width="1438" alt="Screenshot 2022-01-28 at 21 24 30" src="https://user-images.githubusercontent.com/30042483/151643697-26ee4f29-c39b-4adc-923c-581679b0df66.png">
:

I can consistently replicate this issue. It happens on any site that interacts with MetaMask. Is there a fix or a workaround for this?

### Steps to reproduce

Setup:

- Mac OS - Tried on many different machines with different OS versions e.g. 12.1
- Chrome Browser - Different versions e.g. Version 97.0.4692.99 (Official Build) (x86_64)

1. Close Chrome
2. Re-open Chrome
3. Once the browser has opened, go to any site/dapp that interacts with MetaMask e.g. https://nftcontracts.art/arcadelife
4. Check the developer console for error messages

### Error messages or log output

```shell
Unchecked runtime.lastError: Could not establish connection. Receiving end does not exist.
contentscript.js:1 MetaMask: Content script lost connection to "MetaMask Background Multiplex". undefined
contentscript.js:1 MetaMask: Muxed traffic for channel "metamask-provider" failed. undefined
contentscript.js:1 MetaMask: Content script lost connection to "MetaMask Background Legacy Multiplex". undefined
contentscript.js:1 MetaMask: Muxed traffic between channels "provider" and "metamask-provider" failed. undefined
contentscript.js:1 MetaMask: Muxed traffic for channel "publicConfig" failed. undefined
```

### Version

10.8.1

### Browser

Chrome

### Operating system

MacOS

**Comment by kevinghim** [comment]
@hamzasharif tried reproducing using the same OS/Browser combo, and not getting the same errors. Can you provide more details.

**Comment by j-finance** [comment]
Related? https://github.com/MetaMask/metamask-extension/issues/9407

**Comment by hamzasharif** [comment]
> Related? #9407

Yes it is, I can replicate this issue on multiple machines. Each with different OS versions (Mac OS). What kind of info would help with replicating this?

1. Close Chrome completely
2. Open Chrome
3. Navigate to a DApp (an example in the issue description)
4. Try to interact with web3 components etc. After doing so you’ll notice an error in the dev console

Thank you!

**Comment by monsterbitar** [comment]
I've replicated this in firefox on my end. The problem is timing based, if I open my browser and wait for 30s before I go to the first page, I do not get this issue.

My best guess is that there's some point during plugin initialization where if a request is asked of the plugin it has an error due to not being properly initialized but catches that error and prints to the console instead of letting the error propagate.

One solution would be to provide a new function `isReady()` that provides a promise that resolves only when metamask has finished initializing and is ready to be used, such that web app developers can wait for that promise to resolve before making any requets to metamask.

**Comment by erik0612** [comment]
Hi Everyone,
Is there any solution for this problem. I tried to reload app until it's initialized. it's working on chrome and firefox. but not working on metamask browser on the phone.

**Comment by Ravenink** [comment]
It happens to me too. I would love to have a solution.

**Comment by Kstar0722** [comment]
I found temporary solution. the solution is just reload app until `ethereum._state.initialized` is `true`

**Comment by erik0612** [comment]
> I found temporary solution. the solution is just reload app until `ethereum._state.initialized` is `true`

yeah, it works for me as well.

**Comment by lobotomoe** [comment]
Any update on this?

**Comment by bbelly** [comment]
Any update on this?

**Comment by okfirelee** [comment]
ethereum._state.initialized is a totally temp solution. but everytimes reload the website, this state "ethereum._state.initialized" is false. This issue cannot to work better. I think this issue will be fix.

**Comment by axell9641** [comment]
Any news?

**Comment by dievardump** [comment]
Hey there, is there any news on this?

Since today, my MetaMask (Firefox extension) shows a completely blank screen (both popup and home.html)

And any tab opened display those errors in the console:

> MetaMask: Content script lost connection to "MetaMask Background Multiplex". undefined
> MetaMask: Muxed traffic for channel "metamask-provider" failed. undefined 
> MetaMask: Content script lost connection to "MetaMask Background Legacy Multiplex". undefined 
> MetaMask: Muxed traffic between channels "provider" and "metamask-provider" failed. undefined 
> MetaMask: Muxed traffic for channel "publicConfig" failed. undefined

**Comment by Dope-Ass-Ninja** [comment]
I am also able to reproduce this issue on Firefox and Chromium based browsers with the MetaMask extension.

When closing out of the browser completely (tabs and all), and reopening directly to a dApp (say you selected the link from discord, or are testing in a local dev environment which opens the browser to that page for you), MetaMask will consistently throw the above errors and will not work until you refresh the page.

This problem exists not only on my dApps, but very popular ones as well such as UniSwap.

This is quite a breaking issue for front-end developers trying to create a smooth experience for users. An option to refresh the page once MetaMask has loaded or have a function IsReady() as someone has suggested above would definitely help with this issue.

**Comment by AgreenCabTree** [comment]
I solved that issue using this code.
var timerID = setInterval(function() {
    if (document. readyState === 'complete') {
        if (ethereum._state.initialized == true) {
            clearInterval(timerID);
        } else {
            location.href = location.href;
        }
    }
}, 1000);
